### DISCUSSION TITLE
Improving render speed & performance

### DISCUSSION THREAD

**Discussion by silent6trinity** [Syslifters/sysreptor#304] [Q&A]
Hey, just a quick question - I was wondering if there's any suggestions that you may have in regards to squeezing some more speed/performance out of the rendering engine when trying to publish/view reports? I typically will allocate as much resources as possible to my VM's hosting the containers, but nothing really seems to "hit the spot", unless my report is barebones.

Thanks!

**Comment by MWedl** [comment]
The main bottleneck for PDF rendering is weasyprint (HTML to PDF library). Unfortunately, weasyprint is single-threaded, and does not utilize multiple CPU cores.

We observed that using large and complex report designs increases rendering times, especially when they contain large multi-page tables. But that is nothing that can be fixed by adding more hardware resources.

**Reply by MWedl** [reply]
In https://github.com/Syslifters/sysreptor/releases/tag/2024.91 we made some changes in the PDF rendering workflow. SysReptor is now able cancel long-running rendering tasks instead of continue running in the background and use server resources. See #344 

Additionally, detailed render timing information is collected and shown in the PDF preview. This allows to more easily identify the slow part. 
![image](https://github.com/user-attachments/assets/6913f1d3-53b3-47de-bd9a-0e458ba5a7f9)

Additionally, here is a tip to improve performance of table rendering in weasyprint: Set `table-layout: fixed` and manually define column widths. This might drastically reduce rendering times for tables, especially when there are long multiline texts in table cells.
```css
table {
  table-layout: fixed;
}
```

```html
<table>
  <thead>
    <tr>
      <!-- Set explicit column widths -->
      <th style="width: 5em">Column 1</th>
      <th style="width: 10em">Column 2</th>
      <th style="width: auto">Column 3</th>
    </tr>
  </thead>
  <tbody>
    ...
  </tbody>
</table>
```
